rust/sbd-client/src/send_buf.rs

```diff
@@ -0,0 +1,159 @@
+    pub origin: tokio::time::Instant,
+    pub limit_rate: u64,
+    pub idle_keepalive_nanos: u64,
+    pub next_send_at: u64,
+    pub last_send: u64,
+}
+
+impl SendBuf {
+    /// construct a new send buf
+    pub fn new(
+        ws: raw_client::WsRawSend,
+        out_buffer_size: usize,
+        limit_rate: u64,
+        idle_keepalive: std::time::Duration,
+        pre_sent_bytes: usize,
+    ) -> Self {
+        let mut this = Self {
+            ws,
+            buf: VecDeque::default(),
+            out_buffer_size,
+            origin: tokio::time::Instant::now(),
+            limit_rate,
+            idle_keepalive_nanos: idle_keepalive.as_nanos() as u64,
+            next_send_at: 0,
+            last_send: 0,
+        };
+
+        let now = this.origin.elapsed().as_nanos() as u64;
+
+        this.next_send_at = std::cmp::max(now, this.next_send_at)
+            + (pre_sent_bytes as u64 * this.limit_rate);
+
+        this
+    }
+
+    /// Close the connection.
+    pub async fn close(&mut self) {
+        self.ws.close().await;
+    }
+
+    /// We received a new rate limit from the server, update our records.
+    pub fn new_rate_limit(&mut self, limit: u64) {
+        if limit < self.limit_rate {
+            // rate limit updates are sent on a best effort,
+            // and there are network timing conditions to worry about.
+            // assume we accidentally sent a message while the new limit
+            // was in effect, and accout for that in a brute-force manner.
+
+            let now = self.origin.elapsed().as_nanos() as u64;
+
+            self.next_send_at = std::cmp::max(now, self.next_send_at)
+                + (MAX_MSG_SIZE as u64 * self.limit_rate);
+        }
+        self.limit_rate = limit;
+    }
+
+    /// If we need to wait before taking the next step, this
+    /// returns how long.
+    pub fn next_step_dur(&self) -> Option<std::time::Duration> {
+        let now = self.origin.elapsed().as_nanos() as u64;
+
+        if now - self.last_send >= self.idle_keepalive_nanos {
+            // we need a keepalive now, don't wait
+            return None;
+        }
+
+        if now < self.next_send_at {
+            let need_keepalive_at =
+                self.idle_keepalive_nanos - (now - self.last_send);
+            let nanos =
+                std::cmp::min(need_keepalive_at, self.next_send_at - now);
+            Some(std::time::Duration::from_nanos(nanos))
+        } else {
+            None
+        }
+    }
+
+    /// Call `next_step_dur()` first. If it returns None, or you
+    /// await the Duration returned, call this function to send
+    /// out the next queued block on the low-level websocket.
+    /// Returns true if it did something, false if it did not.
+    pub async fn write_next_queued(&mut self) -> Result<bool> {
+        let now = self.origin.elapsed().as_nanos() as u64;
+
+        // first check if we need to keepalive
+        if now - self.last_send >= self.idle_keepalive_nanos {
```

Comment:
i read the ```Call `next_step_dur()` first``` as an instruction to the user. the keepalive check is indeed redundant to the one in that function.

i wonder if it could be simplified by having a single async fn that uses an async sleep to wait and send rather than needing to use two functions sequentially as can be seen here: https://github.com/holochain/sbd/pull/2/files#diff-55f2f95932add9c4e180147077386efd04a8e1ef6635e49bb6fcf514351105b5R125-R128

Reply:
These functions are called from two distinct code paths across multiple threads/tasks. First, if we have filled up our send buffer, they will be called in a back pressure loop from the user public send function potentially many concurrent times (the link @steveej provided). Second, from a single write task that manages making sure buffered data gets sent out even if the app isn't doing any follow-on sends.

I decided the best way to do that from the two paths was to have a non-mutable function that would tell you how long to wait before calling the mutable function that may or may not do anything on account of it potentially getting false positives on the wait side.

Let me know if you think there is a clearer way to do this, or a more clear way to document this.